### PR TITLE
Sanity checks for location values

### PR DESCRIPTION
**Pull request by Davo36**

### CODE REVIEW

File: location.go [hunk 1]
```diff
@@ -141,23 +149,39 @@ func newRawLocationData(r *http.Request) *rawLocationData {
...
-	if !ok && fl.Precision != "" {
+	if !ok && fl.Precision == "" {
+		ok = true // This field is optional, so can be left empty
+	} else {
+		// A value has been set.  Check if it's a sane one.
+		if pre < minPrecision || pre > maxPrecision {
+			ok = false
+		}
+	}
+	if !ok {
```
Comment (mjs) [inline location.go]: this function would be a lot cleaner if the validation was moved into a separate method which could do early returns with an appropriate error if a problem was found

Comment (Davo36) [reply location.go]: Yeah I knew it was getting kinda messy.

So do you mean each of those 2 last checks (altitude and precision) validate in a separate little function?

Comment (Davo36) [reply location.go]: I can do this.  Or I can rewrite the ParseFloat function I guess?

`	pre := 0.0
	if fl.Precision != "" {
		// A value has been set.  Check if it's a sane one.
		pre, ok := parseFloat(fl.Precision)
		if !ok || pre < minPrecision || pre > maxPrecision {
			return nil, newClientError("Invalid precision")
		}
	}
`

Comment (Davo36) [reply location.go]: Oh that didn't format well...

Comment (Davo36) [reply location.go]: `	pre := 0.0

	if fl.Precision != "" {
		// A value has been set.  Check if it's a sane one.
		pre, ok := parseFloat(fl.Precision)
		if !ok || pre < minPrecision || pre > maxPrecision {
			return nil, newClientError("Invalid precision")
		}
	}
`

Comment (Davo36) [reply location.go]: And to help the user I could do this I suppose:

`return nil, newClientError(fmt.Sprintf("Invalid altitude. Should be between %2.2f and %2.2f", minAltitude, maxAltitude)
`

**Comment by Davo36** [comment]
I like that function better now, what do you think?

File: location.go (hunk 1 above)
Comment (mjs) [reply location.go]: > ```
> pre := 0.0
> if fl.Precision != "" {
> 	// A value has been set.  Check if it's a sane one.
> 	pre, ok := parseFloat(fl.Precision)
> 	if !ok || pre < minPrecision || pre > maxPrecision {
> 		return nil, newClientError("Invalid precision")
> 	}
> }
> ```

This is good, but do this in a separate validation method which called from the top of `locationData`.

**Comment by mjs** [comment]
> I like that function better now, what do you think?

Nice. Definitely much better now. I think you might find that it's even better still if all the validation was in a separate method? This would just return `error`.

**Comment by Davo36** [comment]
Well, the validation code in that function creates `newClientError` structs.  With an error that depends on which field is not valid.

This is a type in clienterror.go:

`// clientError represents an error that is caused by the HTTP client
// at the other end of a request.
type clientError struct {
	msg string
}
`

**Comment by mjs** [comment]
> Well, the validation code in that function creates `newClientError` structs. With an error that depends on which field is not valid.

Yep I see that. All I'm suggesting is a small improvement where the validation checks at the top of `locationData()` get moved to a separate validation method which returns `error` (`clientError` matches the `error` interface so can be returned as `error`). This new method would then be called at the top of `locationData()`.

This will tidy up things like not needing to return 2 values (i.e. `nil, some error`) for each validation failure.

**Comment by Davo36** [comment]
Ok, so I can do this:

```
func (fl *rawLocationData) validateLocationData() error {

	lat, ok := parseFloat(fl.Latitude)
	if !ok || lat < -maxLatitude || lat > maxLatitude {
		return newClientError(fmt.Sprintf("Invalid latitude. Should be between %d and %d", -maxLatitude, maxLatitude))
	}
	lon, ok := parseFloat(fl.Longitude)
	if !ok || lon < -maxLongitude || lon > maxLongitude {
		return newClientError(fmt.Sprintf("Invalid longitude. Should be between %d and %d", -maxLongitude, maxLongitude))
	}
	ts, ok := parseTimestamp(fl.Timestamp)
	if !ok {
		return newClientError("Invalid timestamp")
	}
	alt, ok := parseOptionalFloat(fl.Altitude)
	if !ok || alt < minAltitude || alt > maxAltitude {
		return newClientError(fmt.Sprintf("Invalid altitude. Should be between %d and %d", minAltitude, maxAltitude))
	}
	pre, ok := parseOptionalFloat(fl.Precision)
	if !ok || pre < minPrecision || pre > maxPrecision {
		return newClientError(fmt.Sprintf("Invalid precision. Should be between %2.0f and %2.0f", minPrecision, maxPrecision))
	}

	return nil
}

    func (fl *rawLocationData) locationData() (*locationData, error) {

	err := fl.validateLocationData()

	return &locationData{
		Latitude:  lat,
		Longitude: lon,
		Timestamp: ts,
		Altitude:  alt,
		Precision: pre,
	}, nil
}
```
But how then do I get the values out of that method and into a locationData struct?  I either need to pass the values back, and put them into the struct in the locationData method, or create the struct in the validation routine (in which case the locationData method does nothing).

So I couldn't quite see how that would hang together?

**Comment by Davo36** [comment]
* By values I meant lat, lon, etc.

**Comment by mjs** [comment]
>     * By values I meant lat, lon, etc.

Never mind me... I'm doing too many things at once and not paying enough attention.

The code is fine as it is. Approving it now.

**Comment by Davo36** [comment]
Ok, cool.  I thought I was missing something obvious, sitting here thinking about how it could possibly be simpler...

I have changed the word Precision to Accuracy in this code.

So tomorrow I'll push again, and do the go-cptv and thermal-recorder stuff again too.

I know you're away tomorrow, so Wednesday.  No rush from my end of course.

**Pull request merged**